File: app/src/main/java/com/album/sample/SimpleAlbumUI.java
package com.album.sample;

import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.AppCompatTextView;
import android.support.v7.widget.ListPopupWindow;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.Toast;

import com.album.AlbumConstant;
import com.album.model.FinderModel;
import com.album.ui.activity.AlbumBaseActivity;
import com.album.ui.adapter.ListPopupWindowAdapter;
import com.album.ui.annotation.PermissionsType;
import com.album.ui.fragment.AlbumFragment;

import java.util.List;

/**
 * by y on 22/08/2017.
 */

public class SimpleAlbumUI extends AlbumBaseActivity implements AdapterView.OnItemClickListener {

    private Toolbar toolbar;
    private AlbumFragment albumFragment;
    private ListPopupWindow listPopupWindow;
    private AppCompatTextView finerName;

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (albumFragment != null) {
            albumFragment.disconnectMediaScanner();
            albumFragment = null;
        }
        if (listPopupWindow != null) {
            listPopupWindow = null;
        }
    }

    @Override
    protected void initCreate(@Nullable Bundle savedInstanceState) {
        if (albumFragment != null) {
            albumFragment = null;
        }
        initFragment();
        initListPopupWindow();
    }

    private void initListPopupWindow() {
        listPopupWindow = new ListPopupWindow(this);
        listPopupWindow.setAnchorView(finerName);
        listPopupWindow.setWidth(albumConfig.getAlbumListPopupWidth());
        listPopupWindow.setHorizontalOffset(albumConfig.getAlbumListPopupHorizontalOffset());
        listPopupWindow.setVerticalOffset(albumConfig.getAlbumListPopupVerticalOffset());
        listPopupWindow.setModal(true);
        listPopupWindow.setOnItemClickListener(this);
    }

    private void initFragment() {
        FragmentManager supportFragmentManager = getSupportFragmentManager();
        Fragment fragment = supportFragmentManager.findFragmentByTag(AlbumFragment.class.getSimpleName());
        if (fragment != null) {
            albumFragment = (AlbumFragment) fragment;
        } else {
            albumFragment = AlbumFragment.newInstance();
        }
        supportFragmentManager
                .beginTransaction()
                .replace(R.id.sample_album_frame, albumFragment, AlbumFragment.class.getSimpleName())
                .commit();
    }

    @Override
    protected void initView() {
        toolbar = (Toolbar) findViewById(R.id.sample_toolbar);
        finerName = (AppCompatTextView) findViewById(R.id.sample_finder_name);
        finerName.setText(TextUtils.isEmpty(finderName) ? getString(R.string.album_all) : finderName);
        finerName.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                List<FinderModel> finderModel = albumFragment.getFinderModel();
                if (finderModel != null && !finderModel.isEmpty()) {
                    listPopupWindow.setAdapter(new ListPopupWindowAdapter(finderModel));
                    listPopupWindow.show();
                }
            }
        });
    }

    @Override
    protected void initTitle() {
        toolbar.setTitle("sample UI");
        Drawable drawable = ContextCompat.getDrawable(this, albumConfig.getAlbumToolbarIcon());
        drawable.setColorFilter(ContextCompat.getColor(this, albumConfig.getAlbumToolbarIconColor()), PorterDuff.Mode.SRC_ATOP);
        toolbar.setTitleTextColor(ContextCompat.getColor(this, albumConfig.getAlbumToolbarTextColor()));
        toolbar.setNavigationIcon(drawable);
        toolbar.setBackgroundColor(ContextCompat.getColor(this, R.color.colorPrimaryDark));
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_simple_album;
    }

    @Override
    protected void permissionsDenied(@PermissionsType int type) {
        Toast.makeText(this, "permissions error", Toast.LENGTH_SHORT).show();
    }

    @Override
    protected void permissionsGranted(@PermissionsType int type) {
        switch (type) {
            case AlbumConstant.TYPE_PERMISSIONS_ALBUM:
                albumFragment.onScanAlbum(null, false, false);
                break;
            case AlbumConstant.TYPE_PERMISSIONS_CAMERA:
                albumFragment.openCamera();
                break;
        }
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        ListView listView = listPopupWindow.getListView();
        if (listView == null) {
            return;
        }
        ListPopupWindowAdapter adapter = (ListPopupWindowAdapter) listView.getAdapter();
        FinderModel finder = adapter.getFinder(position);
        if (finder == null) {
            return;
        }
        finderName = finder.getDirName();
        finerName.setText(finder.getDirName());
        albumFragment.onScanAlbum(finder.getBucketId(), true, false);
        listPopupWindow.dismiss();
    }

}

File: app/src/main/java/com/album/sample/ui/SimpleAlbumUI.kt
package com.album.sample.ui

import android.graphics.PorterDuff
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.AdapterView
import androidx.appcompat.widget.ListPopupWindow
import androidx.core.content.ContextCompat
import com.album.*
import com.album.sample.R
import com.album.ui.AlbumUiBundle
import com.album.ui.activity.AlbumBaseActivity
import com.album.ui.adapter.FinderAdapter
import com.album.ui.fragment.AlbumFragment
import kotlinx.android.synthetic.main.activity_simple_album.*

/**
 * by y on 22/08/2017.
 */

class SimpleAlbumUI : AlbumBaseActivity(), AdapterView.OnItemClickListener, AlbumParentListener {

    private lateinit var albumFragment: AlbumFragment
    private lateinit var listPopupWindow: ListPopupWindow
    private lateinit var finderAdapter: FinderAdapter

    private lateinit var albumBundle: AlbumBundle
    private lateinit var albumUiBundle: AlbumUiBundle

    override fun initView() {
        listPopupWindow = ListPopupWindow(this)
        sample_finder_name.setOnClickListener {
            val finderEntity = albumFragment.getFinderEntity()
            if (!finderEntity.isEmpty()) {
                finderAdapter.refreshData(finderEntity)
                listPopupWindow.show()
                listPopupWindow.listView?.setBackgroundColor(ContextCompat.getColor(this, albumUiBundle.listPopupBackground))
                return@setOnClickListener
            }
            Album.instance.albumListener?.onAlbumFinderEmpty()
        }
    }

    override fun initCreate(savedInstanceState: Bundle?) {
        albumBundle = intent.extras?.getParcelable(EXTRA_ALBUM_OPTIONS) ?: AlbumBundle()
        albumUiBundle = intent.extras?.getParcelable(EXTRA_ALBUM_UI_OPTIONS) ?: AlbumUiBundle()

        sample_toolbar.title = "sample UI"
        val drawable = ContextCompat.getDrawable(this, albumUiBundle.toolbarIcon)
        drawable?.setColorFilter(ContextCompat.getColor(this, albumUiBundle.toolbarIconColor), PorterDuff.Mode.SRC_ATOP)
        sample_toolbar.setTitleTextColor(ContextCompat.getColor(this, albumUiBundle.toolbarTextColor))
        sample_toolbar.navigationIcon = drawable
        sample_toolbar.setBackgroundColor(ContextCompat.getColor(this, R.color.colorPrimaryDark))
        sample_toolbar.setNavigationOnClickListener {
            Album.instance.albumListener?.onAlbumActivityFinish()
            finish()
        }
        initFragment()
        sample_finder_name.text = if (TextUtils.isEmpty(albumFragment.finderName)) getString(R.string.album_all) else albumFragment.finderName
        initListPopupWindow()
    }

    private fun initFragment() {
        val supportFragmentManager = supportFragmentManager
        val fragment = supportFragmentManager.findFragmentByTag(AlbumFragment::class.java.simpleName)
        if (fragment != null) {
            albumFragment = fragment as AlbumFragment
            supportFragmentManager.beginTransaction().show(fragment).commit()
        } else {
            albumFragment = AlbumFragment.newInstance(albumBundle)
            supportFragmentManager
                    .beginTransaction()
                    .apply { add(R.id.sample_album_frame, albumFragment, AlbumFragment::class.java.simpleName) }
                    .commit()
        }
        albumFragment.albumParentListener = this
    }

    private fun initListPopupWindow() {
        listPopupWindow.anchorView = sample_finder_name
        listPopupWindow.width = albumUiBundle.listPopupWidth
        listPopupWindow.horizontalOffset = albumUiBundle.listPopupHorizontalOffset
        listPopupWindow.verticalOffset = albumUiBundle.listPopupVerticalOffset
        listPopupWindow.isModal = true
        listPopupWindow.setOnItemClickListener(this)
        finderAdapter = FinderAdapter(ArrayList(), albumUiBundle)
        listPopupWindow.setAdapter(finderAdapter)
    }

    override fun getLayoutId(): Int = R.layout.activity_simple_album

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        val finder = finderAdapter.getFinder(position)
        albumFragment.finderName = finder.dirName
        sample_finder_name.text = finder.dirName
        albumFragment.onScanAlbum(finder.bucketId, true, false)
        listPopupWindow.dismiss()
    }

    override fun onAlbumItemClick(multiplePreviewList: ArrayList<AlbumEntity>, position: Int, bucketId: String) {
        SimplePreviewUI.start(albumBundle, albumUiBundle, multiplePreviewList, if (TextUtils.isEmpty(bucketId) && !albumBundle.hideCamera) position - 1 else position, bucketId, albumFragment)
    }

    override fun onBackPressed() {
        Album.instance.albumListener?.onAlbumActivityBackPressed()
        super.onBackPressed()
    }

    override fun onDestroy() {
        albumFragment.disconnectMediaScanner()
        super.onDestroy()
    }

}
